File: listios/StorageStarted/pan3d/display/particle/bone/ParticleBoneData.m
//
//  ParticleBoneData.m
//  StorageStarted
//
//  Created by pan3dme on 2021/6/1.
//  Copyright © 2021 cuiyiran. All rights reserved.
//

#import "ParticleBoneData.h"
#import "MeshData.h"
#import "AnimData.h"
#import "BaseRes.h"
#import "Scene3D.h"
#import "Display3DBoneShader.h"
#import "Display3DBonePartilce.h"
#import "ProgrmaManager.h"

@interface ParticleBoneData ()





@end

@implementation ParticleBoneData





- (void)setAllByteInfo:(ByteArray *)byte
{
    ParticleBoneData* this=self;
    this.meshData = [[MeshData alloc]init:this.scene3D];
    this.animData = [[AnimData alloc]init:this.scene3D];
    this.objScale = [byte readFloat];
    
    int dataWidth = 13;
    int vLen =  [byte getInt];
    int buffStride=dataWidth * 4;
    
  
    
    NSMutableData *dataBase = [[NSMutableData alloc] initWithLength:vLen *buffStride];
    
    this.meshData.vertices=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:3 offset:0 stride:buffStride readType:0];
    this.meshData.uvs=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:2 offset:3 stride:buffStride readType:0];
 
    this.meshData.indexs=   [BaseRes readIntForTwoByte:byte nsdata:dataBase];
    
//    [this upGpuIndexBuffer];
    
    this.meshData.boneIDAry=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:4 offset:5 stride:buffStride readType:2];
    this.meshData.boneWeightAry=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:4 offset:9 stride:buffStride readType:3];
 
    this.meshData.stride =buffStride;
    
    [this readFrameQua:byte];
    
    [super setAllByteInfo:byte];

    [this initVcData];
  
    
    this.meshData.mtkindexs= [self.scene3D.context3D changeObjDataIndexToMtkGpu:  this.meshData.indexs];
    this.meshData.mtkvertices= [self.scene3D.context3D changeDataToGupMtkfloat3:  this.meshData.vertices ];
    this.meshData.mtkuvs= [self.scene3D.context3D changeDataToGupMtkfloat2:  this.meshData.uvs ];
    this.meshData.mtkboneId= [self.scene3D.context3D changeDataToGupMtkfloat4:  this.meshData.boneIDAry ];
    this.meshData.mtkboneWeight= [self.scene3D.context3D changeDataToGupMtkfloat4:  this.meshData.boneWeightAry ];
    this.meshData.mtkuvs=[self.scene3D.context3D changeDataToGupMtkfloat2:  this.meshData.uvs];
    this.meshData.mtkindexCount =   this.meshData.indexs.count;
}
-(void)upGpuIndexBuffer
{
    unsigned int Indices[self.meshData.indexs.count];
    for (int i=0; i<self.meshData.indexs.count; i++) {
        Indices[i]=[self.meshData.indexs[i] intValue];
    }
    
    
    self.meshData.mtkindexs = [self.scene3D.mtkView.device newBufferWithBytes:Indices
                                                     length:sizeof(Indices)
                                                    options:MTLResourceStorageModeShared];
    
    
    self.meshData.trinum=(int)self.objData.indexs.count;
    
    self.meshData.mtkindexCount = self.objData.trinum;
    
 
}
-(void) initVcData {
     
}

-(void) readFrameQua:(ByteArray *)byte
{
    float tempNum = [byte readFloat];
    float RGB32767 = 32767.0;
    int  frameNum    = [byte readInt];
    
    NSMutableArray<DualQuatFloat32Array*>* frameDualQuat = [[NSMutableArray alloc]init];
    for (int i   = 0; i < frameNum; i++) {
        int len = [byte readInt];
        DualQuatFloat32Array* tempDq = [[DualQuatFloat32Array alloc] init];
        GLfloat quat[len * 4];
        GLfloat pos[len * 3];
        for (int j = 0; j < len; j++) {
            quat[j * 4 + 0] = [byte readShort]/RGB32767;
            quat[j * 4 + 1] =  [byte readShort]/RGB32767;
            quat[j * 4 + 2] =  [byte readShort]/RGB32767;
            quat[j * 4 + 3] = [byte readShort]/RGB32767;
            
            pos[j * 3 + 0] =  [byte readShort]/RGB32767* tempNum;
            pos[j * 3 + 1] =  [byte readShort]/RGB32767* tempNum;
            pos[j * 3 + 2] =  [byte readShort]/RGB32767* tempNum;
        }
        NSMutableArray<NSNumber*>* quatArr=[[NSMutableArray alloc]init];
        for(int m=0;m<len * 4;m++){
            [quatArr addObject:[NSNumber numberWithFloat:quat[m]]];
        }
        tempDq.quatArr=[[NSArray alloc]initWithArray:quatArr];
        
        NSMutableArray<NSNumber*>* posArr=[[NSMutableArray alloc]init];
        for(int n=0;n<len* 3;n++){
            [posArr addObject:[NSNumber numberWithFloat:pos[n]]];
        }
        tempDq.posArr=[[NSArray alloc]initWithArray:posArr];
        tempDq.mtkquatArr=  [self.scene3D.context3D changeDataToGupMtkfloat4: tempDq.quatArr];
        tempDq.mtkposArr=  [self.scene3D.context3D changeDataToGupMtkfloat3: tempDq.posArr];
        
    
        
        [frameDualQuat addObject:tempDq];
        
    }
    self.animData.boneQPAry = [[NSMutableArray alloc]init];
    [self.animData.boneQPAry addObject:frameDualQuat];
}
- (Display3DParticle *)getParticle
{
    return [[Display3DBonePartilce alloc]init:self.scene3D];
}
-(void)regShader;
{

   
    if (!self.materialParam) {
          return;
      }
     NSArray<NSNumber*>* shaderParameAry = [self getShaderParam];
    self.materialParam.shader=  [self.scene3D.progrmaManager getMaterialProgram:Display3DBoneShader.shaderStr shaderCls: [[Display3DBoneShader alloc]init:self.scene3D]  material:self.materialParam.material paramAry:shaderParameAry parmaByFragmet:NO];

   
    
}
-(NSArray<NSNumber*>*)getShaderParam;
{
  
    NSArray<NSNumber*>* shaderParameAry = [[NSArray alloc] initWithObjects:@1,@1,@1,@1,@1, nil];
     
    return shaderParameAry;
}

@end

File: android/app/src/main/java/com/pan3d/display/particle/bone/ParticleBoneData.java
package com.pan3d.display.particle.bone;

import com.pan3d.base.ByteArray;
import com.pan3d.base.MeshData;
import com.pan3d.display.particle.Display3DParticle;
import com.pan3d.display.particle.ParticleData;

import com.pan3d.res.BaseRes;
import com.pan3d.scene.Scene3D;
import com.pan3d.vo.AnimData;
import com.pan3d.vo.DualQuatFloat32Array;


import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

public class ParticleBoneData extends ParticleData {
    public MeshData meshData;
    public AnimData animData;
    public float objScale;
    public  ParticleBoneData(Scene3D val  ){
        super(val);
    }

    @Override
    public Display3DParticle getParticle() {
        return  new Display3DBonePartilce(this.scene3D);
    }

    public void setAllByteInfo(ByteArray $byte) {
        this.meshData = new MeshData(this.scene3D);
        this.meshData.indexs=new ArrayList<>();

        this.animData = new AnimData();
        this.objScale = $byte.readFloat();



        int dataWidth = 13;
        int len = $byte.getInt();
        len *= dataWidth * 4;

        byte[] arybuff = new byte[(int)len] ;
        ByteBuffer data=ByteBuffer.wrap(arybuff);

        this.meshData.verticeslist=   BaseRes.readBytes2ArrayBuffer($byte, data, 3, 0, dataWidth, 0);
        this.meshData.uvlist=    BaseRes.readBytes2ArrayBuffer($byte, data, 2, 3, dataWidth, 0);
        BaseRes.readIntForTwoByte($byte,  this.meshData.indexs);
        this.meshData.boneIDAry=   BaseRes.readBytes2ArrayBuffer($byte, data, 4, 5, dataWidth, 2);
        this.meshData.boneWeightAry=    BaseRes.readBytes2ArrayBuffer($byte, data, 4, 6, dataWidth, 3);
        this.meshData.stride = dataWidth * 4;

        this.readFrameQua($byte);
        super.setAllByteInfo($byte);

        this.initVcData();
        this.uploadGpu(this.meshData);
    }
    private void uploadGpu(MeshData val) {
        val.vertexBuffer= val.upGpuvertexBuffer(val.verticeslist);
        val.uvBuffer=val.upGpuvertexBuffer(val.uvlist);
        val.boneIdBuffer=val.upGpuvertexBuffer(val.boneIDAry);
        val.boneWeightBuffer=val.upGpuvertexBuffer(val.boneWeightAry);
        val.indexBuffer= val.upGpuIndexBuffer(val.indexs);
        val.treNum=val.indexs.size();
    }
    public void initVcData() {
//        this.vcmatData = new Float32Array(Display3DBoneShader.getVcSize() * 16);
    }
    private void readFrameQua(ByteArray $byte)    {
        float tempNum  = $byte.readFloat();
        float RGB32767 = 32767;
        int frameNum = $byte.readInt();
        List<DualQuatFloat32Array> $frameDualQuat = new ArrayList<>();
        for (int i = 0; i < frameNum; i++) {
            int $len = $byte.readInt();
            float quat[]=new float[$len * 4];
            float pos[]=new float[$len* 3];
            for (int j = 0; j < $len; j++) {
                quat[j * 4 + 0] = $byte.readShort() / RGB32767;
                quat[j * 4 + 1] = $byte.readShort() / RGB32767;
                quat[j * 4 + 2] = $byte.readShort() / RGB32767;
                quat[j * 4 + 3] = $byte.readShort() / RGB32767;

                pos[j * 3 + 0] = $byte.readShort() / RGB32767 * tempNum;
                pos[j * 3 + 1] = $byte.readShort() / RGB32767 * tempNum;
                pos[j * 3 + 2] = $byte.readShort() / RGB32767 * tempNum;
            }
            DualQuatFloat32Array $tempDq = new DualQuatFloat32Array();
            $tempDq.quatArr=new ArrayList<>();
            $tempDq.posArr=new ArrayList<>();
            for(int m=0;m<quat.length;m++){
                $tempDq.quatArr.add(quat[m]);
            }
            for(int n=0;n<pos.length;n++){
                $tempDq.posArr.add(pos[n]);
            }
            $frameDualQuat.add($tempDq);
        }
        this.animData.boneQPAry = new ArrayList<>();
        this.animData.boneQPAry.add($frameDualQuat);
    }
    @Override
    protected void regShader() {

        if (this.materialParam==null) {
            return;
        }
        List<Boolean>  shaderParameAry =this.getShaderParam();
        this.materialParam.shader3D=    scene3D.progrmaManager.getMaterialProgram(Display3DBoneShader.shaderNameStr,new Display3DBoneShader(scene3D),this.materialParam.material,shaderParameAry,false);
    }

    private List<Boolean> getShaderParam() {
        List<Boolean> shaderParameAry=new ArrayList<>();
        shaderParameAry.add(true );
        shaderParameAry.add(true );
        shaderParameAry.add(true );
        shaderParameAry.add(true );
        return shaderParameAry;

    }

}
